### PR TITLE
[DEP] Use ruff as a linter and formater instead of flake8, isort and black

### PR DESCRIPTION
**Pull request by Mildophin**
### Description

This PR implements ruff as the new formatter and linter for pasqal-cloud, introducing new rules to make the code cleaner.

All the rules I added in the pyproject.toml can be found on: https://docs.astral.sh/ruff/rules/

### Checklist

- [x] The title of the PR follows the right format: [{Label}] {Short Message}. Label examples: IMPROVEMENT, FIX, REFACTORING... Short message is about what your PR changes.

#### Versioning (PASQAL developers only)

- [x] Update the version of pasqal-cloud in `_version.py` following the changes in your PR and by using [semantic versioning](https://semver.org/).

#### Documentation

- [x] Update CHANGELOG.md with a description explaining briefly the changes to the users.

#### Tests

- [x] Unit tests have been added or adjusted.
- [x] Tests were run locally.

#### Internal tests pipeline (PASQAL developers only)
- [ ] Update and run the internal tests while targeting the branch of this PR.
 If your PR hasn't changed any functionality, it still needs to be validated against internal tests.

#### After updating the version (PASQAL developers only)

- [ ] Open a PR on the internal tests that updates the version used for the pasqal-cloud backward compatibility tests.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -46,12 +46,13 @@ def mock_result_link_response():
...
 
 
-def mock_response(request, context):
+def mock_response(request, context):  # noqa: ARG001
     """This acts as a Router to Mock the requests we make with custom behaviors."""
     if request.url.startswith(Endpoints.core):
         return mock_core_response(request)
     if request.url.startswith(RESULT_LINK_ENDPOINT):
         return mock_result_link_response()
+    return None
```
Comment (Mildophin) [inline tests/conftest.py:55]: Why is it here ? Reason: -> https://docs.astral.sh/ruff/rules/implicit-return/

File: pyproject.toml [hunk 2]
```diff
@@ -1,12 +1,61 @@
...
+target-version = "py38"
+
+[tool.ruff.lint]
+# Rules applied by ruff, more information on https://docs.astral.sh/ruff/rules/
+select = [
+    "F",
+    "E",
+    "W",
+    "C90",
+    "F",
```
Comment (nmenetrier) [inline pyproject.toml:25]: You already added Pyflakes in the first line

File: tests/test_batch.py [hunk 3]
```diff
@@ -20,13 +20,13 @@
...
 )
-from tests.test_doubles.authentication import FakeAuth0AuthenticationSuccess
 from tests.conftest import mock_core_response
+from tests.test_doubles.authentication import FakeAuth0AuthenticationSuccess
 
 
 class TestBatch:
     @pytest.fixture(autouse=True)
-    def mock_sleep(self):
+    def _mock_sleep(self):
```
Comment (nmenetrier) [inline tests/test_batch.py:29]: `ruff rule PT004`:

> By convention, fixtures that don't return a value should be named with a
leading underscore, while fixtures that do return a value should not.

Didn't know

**Review by nmenetrier — APPROVED** [review]

File: tests/conftest.py [hunk 4]
```diff
@@ -46,12 +46,13 @@ def mock_result_link_response():
     return {"some": "result"}
 
 
-def mock_response(request, context):
+def mock_response(request, context):  # noqa: ARG001
```
Comment (MohamedAminMallek) [inline tests/conftest.py:49]: I am wondering if doing this helps us avoid  `# noqa: ARG001`

```suggestion
def mock_response(request, _):  
```

File: pasqal_cloud/errors.py [hunk 5]
```diff
@@ -20,7 +20,7 @@ def __init__(self, msg: str, e: Optional[HTTPError] = None) -> None:
         else:
             description = data.get("message", "")
 
-        super().__init__(
+        super().__init__(  # noqa: RET503
```
Comment (MohamedAminMallek) [inline pasqal_cloud/errors.py:23]: Why not just add the return statement?
```suggestion
        return super().__init__( 
```

File: tests/conftest.py [hunk 6]
```diff
@@ -107,39 +108,38 @@ def request_mock_exception_batch_creation(
     return mock
 
 
-@pytest.fixture(scope="function")
+@pytest.fixture()
```
Comment (MohamedAminMallek) [inline tests/conftest.py:111]: Is this change related to ruff?

**Review by MohamedAminMallek — APPROVED** [review]

File: tests/conftest.py (hunk 6 above)
Comment (Mildophin) [reply tests/conftest.py:106]: yes, PT family rule: https://docs.astral.sh/ruff/rules/#flake8-pytest-style-pt

File: pasqal_cloud/client.py [hunk 7]
```diff
@@ -117,25 +117,25 @@ def _request(
...
     def _send_batch(self, batch_data: Dict[str, Any]) -> Dict[str, Any]:
         batch_data.update({"project_id": self.project_id})
-        batch_data = self._request(
+        response: Dict[str, Any] = self._request(
             "POST",
             f"{self.endpoints.core}/api/v1/batches",
             batch_data,
         )["data"]
-        return batch_data
+        return response
```
[the remark this answers is not in the document]
Comment (Mildophin) [reply pasqal_cloud/client.py:159]: Made by myself, but it was suggested implicitly by mypy actually as I had errors related to it.

File: pasqal_cloud/batch.py [hunk 8]
```diff
@@ -162,7 +163,7 @@ def add_jobs(
 
         if wait:
             while any(
-                [job.status in {"PENDING", "RUNNING"} for job in self.ordered_jobs]
+                job.status in {"PENDING", "RUNNING"} for job in self.ordered_jobs
```
[the remark this answers is not in the document]
Comment (Mildophin) [reply pasqal_cloud/batch.py:165]: Added it, It is linked to the C4 family rule: https://docs.astral.sh/ruff/rules/#flake8-comprehensions-c4

File: .github/workflows/run-tests-and-mypy-and-pre-commit-hooks.yml [hunk 9]
```diff
@@ -34,7 +34,7 @@ jobs:
 
       - name: Perform unit tests
         run: |
-          pytest -vvv --cov pasqal_cloud tests
```
Comment (Mildophin) [inline .github/workflows/run-tests-and-mypy-and-pre-commit-hooks.yml:37]: Was not needed, it could avoid launching some tests if outside the folders.

**Review by MatthieuMoreau0 — COMMENTED** [review]
Looks ok to me

**Review by nmenetrier — APPROVED** [review]
